src/main/scala/mesosphere/mesos/ResourceMatcher.scala

```diff
@@ -20,20 +36,40 @@ object ResourceMatcher {
   def matchResources(offer: Offer, app: AppDefinition, runningTasks: => Set[MarathonTask],
                      acceptedResourceRoles: Set[String] = Set("*")): Option[ResourceMatch] = {
 
-    val groupedResources = offer.getResourcesList.asScala.groupBy(_.getName)
+    val groupedResources: Map[Role, mutable.Buffer[Protos.Resource]] = offer.getResourcesList.asScala.groupBy(_.getName)
 
-    def findScalarResourceRole(tpe: String, value: Double): Option[Role] =
-      groupedResources.get(tpe).flatMap {
-        _
-          .filter(resource => acceptedResourceRoles(resource.getRole))
-          .find { resource =>
-            resource.getScalar.getValue >= value
-          }.map(_.getRole)
+    def findScalarResource(name: String, requiredValue: Double): Option[ScalarMatch] =
+      groupedResources.get(name).flatMap { resources =>
+        val matchingScalarResources = resources.filter { resource =>
+          acceptedResourceRoles(resource.getRole) && resource.hasScalar
+        }
+
+        val asMatches = matchingScalarResources.map { resource =>
+          ScalarMatch(
+            requiredValue = requiredValue,
+            offeredValue = resource.getScalar.getValue,
+            role = resource.getRole
+          )
+        }
+
+        if (asMatches.isEmpty) {
+          None
+        }
+        else {
+          Some(asMatches.maxBy(_.offeredValue))
+        }
       }
 
-    def cpuRoleOpt: Option[Role] = findScalarResourceRole(Resource.CPUS, app.cpus)
-    def memRoleOpt: Option[Role] = findScalarResourceRole(Resource.MEM, app.mem)
-    def diskRoleOpt: Option[Role] = findScalarResourceRole(Resource.DISK, app.disk)
+    val cpuRoleOpt: Option[ScalarMatch] = findScalarResource(Resource.CPUS, app.cpus)
```

Comment:
I think naming these *RoleOpt is really confusing, since they are options of ScalarMatch.